src/cluster/src/check/mod.rs

```diff
@@ -379,16 +398,208 @@ impl ClusterChecker {
+                        Err(e) => {
+                            // If the fix failed, wrap the original failed check in Unrecoverable
+                            results.push(Ok(CheckStatus::fail(CheckFailed::Unrecoverable(e))));
+                            // We return upon the first check failure
+                            return results;
+                        }
+                    }
+                }
+                it @ Ok(CheckStatus::Fail(_)) => {
+                    results.push(it);
+                    return results;
+                }
+                it @ Err(_) => {
+                    results.push(it);
+                    return results;
+                }
+            }
         }
 
-        CheckResults::from(results)
+        results
+    }
+
+    /// Performs all checks in an async task, returning the results via a channel.
+    ///
+    /// This function will return immediately with a channel which will yield progress
+    /// updates about checks as they are run.
+    ///
+    /// If you want to run the checks as a single batch and receive all of the results
+    /// at once, use [`run`] instead.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use fluvio_cluster::ClusterChecker;
+    /// # async fn do_run_with_progress() {
+    /// use async_channel::Receiver;
+    /// let progress: Receiver<CheckResult> = ClusterChecker::empty()
+    ///     .with_preflight_checks()
+    ///     .run_with_progress();
+    /// while let Ok(check_result) = progress.recv().await {
+    ///     println!("Got check result: {:?}", check_result);
+    /// }
+    /// # }
+    /// ```
+    pub fn run_with_progress(self) -> Receiver<CheckResult> {
+        let (sender, receiver) = async_channel::bounded(100);
+        spawn(async move {
+            for check in self.checks {
+                let check_result = check.perform_check().await;
+
+                // Nothing we can do if channel fails
+                let progress_result = sender.send(check_result).await;
+
+                // If channel fails, the best we can do is log it
+                if let Err(error) = progress_result {
+                    warn!(%error, "Failed to send check progress update to client:");
+                }
+            }
+        });
+        receiver
+    }
+
+    /// Performs all checks in an async task and attempts to fix anything it can.
+    ///
+    /// This function will return immediately with a channel which will yield
+    /// progress updates about checks and fixes as they run.
+    ///
+    /// If you want to run checks and fixes as a single batch and receive all of
+    /// the results at once, use [`run`] instead.
+    pub fn run_and_fix_with_progress<F, R>(self, fix: F) -> Receiver<CheckResult>
+    where
+        F: Fn(RecoverableCheck) -> R + Send + Sync + 'static,
+        R: Future<Output = Result<(), UnrecoverableCheck>> + Send + Sync,
```

Comment:
this seems to be duplicate of `run_and_fix`.   this should be consolidated into a single method with only difference being what to do with result in between

Reply:
The way I see it, there are two orthogonal concerns:

- Just Check
  - Batch (`run`)
  - Progress (`run_with_progress`)
- Check and Fix
  - Batch (`run_and_fix`)
  - Progress (`run_and_fix_with_progress`)

So the two concerns are a) whether to fix the checks and b) whether to report progress as they go.

I wanted to solve each of these generally so that it would work for either a local install or a k8 install.

For a), I abstracted away the idea of a fixing-function. So the caller of the checker will provide their own function that describes how to take a "failed" `RecoverableCheck` and attempt to fix it. That's the `fix: F` function argument. All of the generics basically say that the fix function may be an async function.

For b), I wanted to maintain the library/binary separation while also providing a way for the binary half (i.e. the CLI) to be receiving updates from checks being performed by code in the library half. To do this, I made it so that the `with_progress` functions immediately return with an `async_channel::Receiver<CheckResult>`. Inside the `with_progress` function, it spawns a task which is what actually performs the checks, and it sends the results through the channel as it goes. So the CLI can simply watch for new `CheckResult`s to come through the channel and print them to stdout.